wait for serialized clientId when loading with stashed ops

wes-carlson opened the pull request:
close #9240
Wait for a leave message from the previously serialized container's clientId before transitioning to "connected" state when loading with stashed ops.

In #9242 we will no longer see the leave op, so we will need to cancel the timer at "connect" event if the serialized clientId is not in the quorum.

wes-carlson committed:
wait for serialized clientId

wes-carlson committed:
remove IPendingLocalState from container-definitions

heliocliu committed:
Merge pull request #9379 from microsoft/main

Merge "main" into "next" in prep for "next" version bump

vladsud commented on packages/loader/container-loader/src/connectionStateHandler.ts:
```diff
@@ -65,6 +67,10 @@ export class ConnectionStateHandler {
             },
         );
+        // if we have a serialized clientId we will need to wait unless we see its "leave" op
+        if (serializedClientId !== undefined) {
```
While connection state manager needs to receive new info (old clientId) in constructor, I believe adding new property makes it more complex and thus more error prone. Constructor should set up object as if we just disconnected (same way as do to right here).
But we should use existing properties only (including this.clientId) and move to (setup) some known existing state that already handles this condition, not create more states and reason about more complex state machine.
In other words, there should not be this.serializedClientId IMHO.

+ @markfields, who is poking at this code right now.

vladsud commented:
It's not enough to have serialized client ID at that level. We have to restore all layers to exactly same state as if no serialization / rehydration happened.

For example, container runtime needs to set clientId to the one we had such that it can reason about incoming ops, if they are "local" and thus have to be acked. Same for Container layer, who (I believe) calculates local flag (though I'm sure that's not the only thing we depend on client Id).

It would be great to see this process unfold in one PR, even if we want to break it into multiple small PRs. And have UTs that test each individual layer, but also end-to-end tests that validate receiving ops from old client after rehydration.

heliocliu commented:
Does this change need to go into the 0.58 release?  If so, please merge it before Wednesday and let me know so we can copy it over.  Otherwise please merge this into the `next` branch instead because it is a breaking change.

anthony-murphy committed:
MergeTree: Unify removed client id and overlap clients (#9369)

This change unifies the removedClientId and removedClientOverlap into a single array, removedClientIds.

This simplifies our handling in a number of places, as we no longer need to check and manage both. In fact, there were a couple bugs where both were not handled appropriately.

In addition, I've improved how we handle IRemovalInfo such that it is now all or nothing, which removes the need for a number of unsafe non-null assertions, and make the code simpler by removing the need to repeated non-null checks.

this may fix #9138, but it hard to know as we don't have much telemetry around the issue.

heliocliu committed:
Integrate main to next (#9408)

* Run static web app deployment in linux, and guardian tasks in windows (#9386)

* Integrate latest prerelease of eslint-config-fluid throughout the repo and fix up errors (#9389)

* Support string parserOptions.project in eslint configuration (#9392)

* Handle errors in docs pipeline more gracefully (#9393)

* Fixing some error logging race conditions for the root datastore end-to-end tests (#9387)

it's disabling the summarizer, which either runs or not, depending on how long the test takes.

* Remove duplicate implementation from toInstrumentedOdspTokenFetcher (toInstrumentedTokenFetcher) (#9375)

* Remove duplicate toInstrumentedTokenFetcher implementation (duplicate from toInstrumentedOdspTokenFetcher )

* MergeTree: Add stress for LocalReferences (#9340)

This change leverages the existing stress infra in merge tree, we call them farms as in bug farm as they find lots of bugs. This is in preparation for how we do local reference sliding.

related to #1008

* Fix lint issues

* Missed one

Co-authored-by: Tyler Butler <tyler@tylerbutler.com>
Co-authored-by: Matt Rakow <ChumpChief@users.noreply.github.com>
Co-authored-by: Andrei Iacob <84357545+andre4i@users.noreply.github.com>
Co-authored-by: NicholasCouri <nichoc@microsoft.com>
Co-authored-by: Tony Murphy <anthony.murphy@microsoft.com>
Co-authored-by: Matt Rakow <marakow@microsoft.com>
